aptos-move/aptos-vm/src/move_vm_ext/write_op_converter.rs

```diff
@@ -30,22 +31,43 @@ macro_rules! convert_impl {
             move_storage_op: MoveStorageOp<Bytes>,
             legacy_creation_as_modification: bool,
         ) -> Result<WriteOp, VMStatus> {
+            let move_storage_op = match move_storage_op {
+                MoveStorageOp::New(data) => MoveStorageOp::New((data, None)),
+                MoveStorageOp::Modify(data) => MoveStorageOp::Modify((data, None)),
+                MoveStorageOp::Delete => MoveStorageOp::Delete,
+            };
             self.convert(
                 self.remote.$get_metadata_callback(state_key),
-                move_storage_op,
+                &move_storage_op,
                 legacy_creation_as_modification,
             )
         }
     };
 }
 
 impl<'r> WriteOpConverter<'r> {
-    convert_impl!(convert_resource, get_resource_state_value_metadata);
-
     convert_impl!(convert_module, get_module_state_value_metadata);
 
     convert_impl!(convert_aggregator, get_aggregator_v1_state_value_metadata);
 
+    pub(crate) fn convert_resource(
+        &self,
+        state_key: &StateKey,
+        move_storage_op: MoveStorageOp<BytesWithLayout>,
+        legacy_creation_as_modification: bool,
+    ) -> Result<(WriteOp, Option<MoveTypeLayout>), VMStatus> {
+        let result = self.convert(
+            self.remote.get_resource_state_value_metadata(state_key),
+            &move_storage_op,
+            legacy_creation_as_modification,
+        );
+        match move_storage_op {
```

Comment:
matching first and then converting would avoid problems with not consuming move_storage_op in convert

Reply:
The match statement uses the result of the first function call. So, unfortunately couldn't have the match statement before.